File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/RuntimeAutomataAstApplication.java
package titan.ast.runtime;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import titan.ast.runtime.AstGeneratorResult.AstResult;
import titan.ast.runtime.AstGeneratorResult.TokensResult;

/**
 * .
 *
 * @author tian wei jun
 */
public class RuntimeAutomataAstApplication implements Cloneable {
  AutomataData automataData;
  TokenAutomata tokenAutomata;
  AstAutomata astAutomata;

  public void setContext(InputStream automataByteInputStream) throws AutomataDataIoException {
    PersistentAutomataData persistentAutomataData =
        new PersistentAutomataData(automataByteInputStream);
    AutomataData automataData = buildAutomataData(persistentAutomataData);
    setContext(automataData);
  }

  public void setContext(String automataFilePath) throws AutomataDataIoException {
    try (InputStream automataByteInputStream = new FileInputStream(automataFilePath)) {
      PersistentAutomataData persistentAutomataData =
          new PersistentAutomataData(automataByteInputStream);
      AutomataData automataData = buildAutomataData(persistentAutomataData);
      setContext(automataData);
    } catch (IOException e) {
      throw new AutomataDataIoException(e);
    }
  }

  public void setContext(AutomataData automataData) {
    this.automataData = automataData;
    tokenAutomata = new TokenAutomataBuilder().build(automataData);
    astAutomata = new AstAutomataBuilder().build(automataData);
  }

  private AutomataData buildAutomataData(PersistentAutomataData persistentAutomataData)
      throws AutomataDataIoException {
    PersistentAutomataObject persistentAutomataObject =
        new PersistentAutomataObject(persistentAutomataData);
    persistentAutomataObject.init();
    return persistentAutomataObject.toAutomataData();
  }

  public AstGrammar[] getGrammars() {
    Grammar[] oriGrammars = automataData.grammars;
    AstGrammar[] grammars = new AstGrammar[oriGrammars.length];
    for (int indexOfGrammar = 0; indexOfGrammar < oriGrammars.length; indexOfGrammar++) {
      Grammar oriGrammar = oriGrammars[indexOfGrammar];
      grammars[indexOfGrammar] = new AstGrammar(oriGrammar.name, oriGrammar.type);
    }
    return grammars;
  }

  public AstGeneratorResult buildAst(String sourceFilePath) {
    TokensResult tokensResult = tokenAutomata.buildToken(sourceFilePath);
    AstResult astResult;
    if (tokensResult.isOk()) {
      astResult = astAutomata.buildAst(tokensResult.getOkData());
    } else {
      astResult = AstResult.generateTokensErrorResult();
    }
    return new AstGeneratorResult(tokensResult, astResult);
  }

  public AstGeneratorResult buildAst(InputStream sourceByteInputStream) {
    TokensResult tokensResult = tokenAutomata.buildToken(sourceByteInputStream);
    AstResult astResult;
    if (tokensResult.isOk()) {
      astResult = astAutomata.buildAst(tokensResult.getOkData());
    } else {
      astResult = AstResult.generateTokensErrorResult();
    }
    return new AstGeneratorResult(tokensResult, astResult);
  }

  public void displayGraphicalViewOfAst(Ast ast) {
    new titan.ast.runtime.AstGuiOutputer(ast).output();
  }

  public void displayGraphicalViewOfAst(Ast ast, String charsetName) {
    new titan.ast.runtime.AstGuiOutputer(ast, charsetName).output();
  }

  @Override
  public RuntimeAutomataAstApplication clone() {
    RuntimeAutomataAstApplication cloner = null;
    try {
      cloner = (RuntimeAutomataAstApplication) super.clone();
    } catch (CloneNotSupportedException e) {
      cloner = new RuntimeAutomataAstApplication();
    }
    cloner.setContext(this.automataData);
    return cloner;
  }
}

File: runtime/cpp/titan-ast-runtime-lib/RuntimeAutomataAstApplication.cpp
//
// Created by tian wei jun on 2022/11/23 0023.
//

#include "RuntimeAutomataAstApplication.h"
#include "AstAutomataBuilder.h"
#include "PersistentData.h"
#include "PersistentObject.h"
#include "TokenAutomataBuilder.h"
#include <iostream>
#include <list>

std::mutex RuntimeAutomataAstApplication::cloneLock{};

RuntimeAutomataAstApplication::RuntimeAutomataAstApplication()
    : automataData(std::make_shared<AutomataData>()), tokenAutomata(nullptr),
      astAutomata(nullptr) {}

RuntimeAutomataAstApplication::~RuntimeAutomataAstApplication() {
  delete tokenAutomata;
  tokenAutomata = nullptr;
  delete astAutomata;
  astAutomata = nullptr;

  cloneLock.lock();
  //引用计数减一
  automataData.reset();
  cloneLock.unlock();
}

BuildAutomataResult
RuntimeAutomataAstApplication::setContext(const std::string *automataFilePath) {
  PersistentData persistentData;
  BuildAutomataResult buildAutomataResult =
      persistentData.init(automataFilePath);
  //初始化错误（可能原因：自动机文件不存在）
  if (!buildAutomataResult.isOk) {
    return buildAutomataResult;
  }
  PersistentObject persistentObject;
  buildAutomataResult = persistentObject.initByPersistentData(&persistentData);
  if (!buildAutomataResult.isOk) {
    return buildAutomataResult;
  }

  auto ptrAutomataData = this->automataData.get();
  // all heap data is moved
  persistentObject.setAutomataData(ptrAutomataData);

  TokenAutomataBuilder tokenAutomataBuilder;
  tokenAutomata = tokenAutomataBuilder.build(ptrAutomataData);

  AstAutomataBuilder astAutomataBuilder;
  astAutomata = astAutomataBuilder.build(ptrAutomataData);
  return {true, ""};
}

AstGeneratorResult *
RuntimeAutomataAstApplication::buildAst(const std::string *sourceCodeFilePath) {
  TokensResult *tokensResult = tokenAutomata->buildToken(sourceCodeFilePath);
  AstResult *astResult = nullptr;
  if (tokensResult->isOk()) {
    astResult = astAutomata->buildAst(tokensResult->getOkData());
  } else {
    astResult = AstResult::generateTokensErrorResult();
  }

  return new AstGeneratorResult(tokensResult, astResult);
}

RuntimeAutomataAstApplication *RuntimeAutomataAstApplication::clone() {
  auto *app = new RuntimeAutomataAstApplication();
  cloneDataToCloner(app);
  return app;
}

void RuntimeAutomataAstApplication::cloneDataToCloner(
    RuntimeAutomataAstApplication *cloner) {
  cloneLock.lock();
  //引用计数增加
  cloner->automataData = this->automataData;
  cloneLock.unlock();

  AutomataData *ptrAutomataData = cloner->automataData.get();

  TokenAutomataBuilder tokenAutomataBuilder;
  cloner->tokenAutomata = tokenAutomataBuilder.build(ptrAutomataData);

  AstAutomataBuilder astAutomataBuilder;
  cloner->astAutomata = astAutomataBuilder.build(ptrAutomataData);
}

std::vector<AstGrammar> RuntimeAutomataAstApplication::getGrammars() {
  std::vector<AstGrammar> grammars(automataData->sizeOfGramamrs);
  for (int indexOfGrammars = 0; indexOfGrammars < automataData->sizeOfGramamrs;
       indexOfGrammars++) {
    auto originalGrammar = automataData->grammars[indexOfGrammars];
    AstGrammar grammar(originalGrammar->name, originalGrammar->type);
    grammars[indexOfGrammars] = grammar;
  }
  return grammars;
}
